src/solver/pooling/backwardNd.cpp

```diff
@@ -54,165 +69,180 @@ PoolingBackwardNd::GetSolution(const ExecutionContext&,
+        KernelBuildParameters{
+            {"MLO_POOLING_OP_ID", pooling_method},
+            {"MAX_ACTIV_WORKITEM", max_activ_workitem},
+            {"MLO_POOLING_GROUP_SZ0", wavesize},
+            {"MLO_POOLING_GROUP_SZ1", 1},
+            {"MLO_POOLING_GROUP_SZ2", 1},
+            {"PIX_W_PER_WORK", pix_w_per_work},
+            {"PIX_H_PER_WORK", pix_h_per_work},
+            {"PIX_D_PER_WORK", pix_d_per_work},
+            {"KERNEL_SZ_D", lens[0]},
+            {"KERNEL_SZ_H", lens[1]},
+            {"KERNEL_SZ_W", lens[2]},
+            {"STRIDE_D", strides[0]},
+            {"STRIDE_H", strides[1]},
+            {"STRIDE_W", strides[2]},
+            {"TERRITORY_OVERLAP", static_cast<int>(territory_overlap)},
+            {"MLO_POOLING_INDEX_TYPE",
+             get_pooling_index_type_name(problem.GetPooling().GetIndexType())},
+            {"MLO_POOLING_INDEX_MAX",
+             get_pooling_index_type_max_name(problem.GetPooling().GetIndexType())},
+        }
+        << GetDataTypeKBP(problem.GetDYDesc().GetType());
 
-            const auto top_d = *(params.dyDesc.GetLengths().rbegin() + 2);
-            const auto top_h = *(params.dyDesc.GetLengths().rbegin() + 1);
-            const auto top_w = *(params.dyDesc.GetLengths().rbegin());
+    kernel.comp_options = build_params.GenerateFor(kbp::OpenCL{});
 
-            int pix_w_per_work = 1;
-            int pix_h_per_work = 4;
-            int pix_d_per_work = 2;
+    kernel.l_wk = {wavesize, 1, 1};
+    kernel.g_wk = {wavesize * grp_num, 1, 1};
 
-            int batch = params.dyDesc.GetLengths()[0];
-            int chal  = params.dyDesc.GetLengths()[1];
+    result.construction_params.push_back(kernel);
 
-            int bot_d = *(params.dxDesc.GetLengths().rbegin() + 2);
-            int bot_h = *(params.dxDesc.GetLengths().rbegin() + 1);
-            int bot_w = *(params.dxDesc.GetLengths().rbegin());
+    const auto top_d = is2d ? 1 : *(top.GetLengths().rbegin() + 2);
+    const auto top_h = *(top.GetLengths().rbegin() + 1);
+    const auto top_w = *(top.GetLengths().rbegin());
 
-            int pix_blk_w = std::max((bot_w + pix_w_per_work - 1) / pix_w_per_work, 1);
-            int pix_blk_h = std::max((bot_h + pix_h_per_work - 1) / pix_h_per_work, 1);
-            int pix_blk_d = std::max((bot_d + pix_d_per_work - 1) / pix_d_per_work, 1);
+    const auto bot_strides  = bot.GetStrides();
+    const auto bot_n_stride = bot_strides[0];
+    const auto bot_c_stride = bot_strides[1];
+    const auto bot_d_stride = bot_strides[2]; // Same as bot_h_stride in 3D converted from 2D.
+    const auto bot_h_stride = is2d ? bot_strides[2] : bot_strides[3];
+    const auto top_strides  = top.GetStrides();
+    const auto top_n_stride = top_strides[0];
+    const auto top_c_stride = top_strides[1];
+    const auto top_d_stride = top_strides[2]; // Same as top_h_stride in 3D converted from 2D.
+    const auto top_h_stride = is2d ? top_strides[2] : top_strides[3];
```

Comment:
I would prefer the variant which is easier to read. Which one is easier for you? just tell me and I'll follow ;)

Reply:
Probably initial version may be a bit easier to read, but it generates a lot of identical copy-pasted code and can be error-prone.
In case of lambda it explicitly states that strides layouts are identical, and it can be modified in an easier and safer way, so you won't miss one index somewhere or get `top` strides from `bot` tensor (except the case where strides layout becomes different and everything will be split anyways).

Also I'd preferred explicit naming for operations.

Reply:
This leads to `error: reference to local binding 'bot_n_stride' declared in enclosing function 'miopen::solver::pooling::PoolingBackwardNd::GetSolution'`, see https://stackoverflow.com/questions/46114214/lambda-implicit-capture-fails-with-variable-declared-from-structured-binding

Reply:
I am going to use std::tie. The constness is lost.

Reply:
@CAHEK7 I recommend looking at the result at 8be604cca0b503d9eeb6a06a710aa0f900401d3f
[resolved]

Reply:
> @CAHEK7 I recommend looking at the result at [8be604c](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/8be604cca0b503d9eeb6a06a710aa0f900401d3f) [resolved]

@atamazov , I agree it looks ugly (

As about the constness it does not matter in this particular context, because later it captured by value and constness can be guaranteed since invoker lambda is non-mutable (or will be lost anyway, if lambda is mutable).

Reply:
[Notice] Possibility of constness can be used as a criterion of good design. If the design does not allow creating const entities wherever those entities can not change, then the design well... er... may be improved.